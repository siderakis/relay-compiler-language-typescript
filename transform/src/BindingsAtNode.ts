import * as ts from 'typescript';
export type Bindings = Map<string, BindingInfo>;
export const enum BindingKind {
  Local,
  Module,
  Import,
  Require,
  Export
}
export interface BindingInfo {
  kind: BindingKind;
}
export function bindingsAtNode(node: ts.Node): Bindings {
  const bindings = new Map<string, BindingInfo>();

  function findInitializerBindings(initializer: ts.ForInitializer) {
    if (ts.isVariableDeclarationList(initializer)) {
      initializer.declarations.forEach(decl => {
        processBinding(decl.name, BindingKind.Local);
      });
    }
  }

  function determineBindingKindFromInitializer(initializer: ts.Expression | undefined): BindingKind {
    if (initializer == null) {
      return BindingKind.Local;
    }
    if (ts.isCallExpression(initializer) && ts.isIdentifier(initializer.expression) && initializer.expression.text === 'require') {
      return BindingKind.Require;
    }
    return BindingKind.Local;
  }

  function findStatementsBindings(block: ts.Block | ts.SourceFile) {
    block.statements.forEach(stmt => {
      if (ts.isClassDeclaration(stmt)) {
        if (stmt.name) {
          addBinding(stmt.name.text, BindingKind.Local);
        }
        return;
      }
      if (ts.isExportDeclaration(stmt)) {
        if (stmt.exportClause) {
          stmt.exportClause.elements.forEach(element => {
            addBinding(element.name.text, BindingKind.Export);
          });
        }
        if (stmt.name) {
          addBinding(stmt.name.text, BindingKind.Export);
        }
        return;
      }
      if (ts.isImportDeclaration(stmt) && stmt.importClause != null) {
        const namedBindings = stmt.importClause.namedBindings;
        if (namedBindings == null) {
          return;
        }
        if (ts.isNamespaceImport(namedBindings)) {
          addBinding(namedBindings.name.text, BindingKind.Import);
        } else {
          namedBindings.elements.forEach(element => {
            addBinding((element.propertyName || element.name).text, BindingKind.Import);
          });
        }
        return;
      }
      if (ts.isVariableStatement(stmt)) {
        stmt.declarationList.declarations.forEach(decl => {
          processBinding(decl.name, determineBindingKindFromInitializer(decl.initializer));
        });
        return;
      }
      // One can declare named functions in tons of weird ways that will end up in scope that this won't catch
      // the same goes for var declarations deep within the code structure.
      if (ts.isExpressionStatement(stmt) && ts.isFunctionExpression(stmt.expression)) {
        if (stmt.expression.name != null) {
          addBinding(stmt.expression.name.text, BindingKind.Local);
        }
        return;
      }
    });
  }

  function addBinding(name: string, kind: BindingKind) {
    if (!bindings.has(name)) {
      bindings.set(name, { kind: kind });
    }
  }

  function processBinding(binding: ts.Identifier | ts.ArrayBindingPattern | ts.ObjectBindingPattern, kind: BindingKind) {
    if (ts.isIdentifier(binding)) {
      addBinding(binding.text, kind);
      return;
    }
    if (ts.isArrayBindingPattern(binding)) {
      binding.elements.forEach(element => {
        if (ts.isBindingElement(element)) {
          processBinding(element.name, kind);
        }
      });
      return;
    }
    binding.elements.forEach(element => {
      if (ts.isBindingElement(element)) {
        processBinding(element.name, kind);
      }
    });
  }

  while (node.parent != null) {
    const parent = node.parent;
    node = parent;
    if (ts.isBlock(parent) || ts.isSourceFile(parent)) {
      findStatementsBindings(parent);
      continue;
    }
    if (ts.isForStatement(parent) && parent.initializer != null) {
      findInitializerBindings(parent.initializer);
      continue;
    }
    if (ts.isForOfStatement(parent)) {
      findInitializerBindings(parent.initializer);
      continue;
    }
  }
  return bindings;
}
