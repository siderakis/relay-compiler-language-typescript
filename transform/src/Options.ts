import { GraphQLSchema, buildClientSchema, buildASTSchema, parse } from "graphql";
import * as fs from 'fs';

export interface Options {
	artifactDirectory?: string;
	compat?: boolean;
	schema?: string;
	isDevVariable?: string;
	buildCommand?: string;
	isDevelopment?: boolean;
}

export interface NormalizedOptions {
	artifactDirectory?: string;
	compat?: boolean;
	schema?: GraphQLSchema;
	isDevVariable?: string;
	buildCommand?: string;
	isDevelopment?: boolean;
}

const dotJsonLength = '.json'.length;
const dotGraphQLLength = '.graphql'.length;

export function readGraphQLSchema(schemaPath: string): GraphQLSchema {
	const contents = fs.readFileSync(schemaPath, { encoding: 'utf8' });
	if (schemaPath.substring(schemaPath.length - dotJsonLength) === '.json') {
		const json = JSON.parse(contents);
		if (json.__schema) {
			return buildClientSchema(json);
		}
		if (json.data && json.data.__schema) {
			return buildClientSchema(json.data);
		}
		throw new Error('Expected data file to contain a JSON encoded GraphQLSchema');
	} else if (schemaPath.substring(schemaPath.length - dotGraphQLLength) === '.graphql') {
		return buildASTSchema(parse(contents));
	}
	throw new Error('Unsupported file. schema option only supports json and graphql file extensions');
}

export function normalizeOptions(options: Options): NormalizedOptions {
	return {
		...options,
		schema: options.schema ? readGraphQLSchema(options.schema) : undefined,
	};
}
