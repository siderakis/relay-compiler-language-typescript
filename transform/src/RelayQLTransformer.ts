import { RelayQLPrinter } from "./RelayQLPrinter";

import * as util from "util";
import * as ts from "typescript";

import {
  RelayQLFragment,
  RelayQLMutation,
  RelayQLQuery,
  RelayQLSubscription,
} from "./RelayQLAST";
import {
  formatError,
  parse,
  Source,
  validate,
  FieldsOnCorrectTypeRule,
  FragmentsOnCompositeTypesRule,
  KnownArgumentNamesRule,
  KnownTypeNamesRule,
  PossibleFragmentSpreadsRule,
  VariablesInAllowedPositionRule,
  ProvidedNonNullArgumentsRule,
} from "graphql";
import * as gql from "graphql";

const ValuesOfCorrectTypeRule: typeof PossibleFragmentSpreadsRule = (gql as any).ValuesOfCorrectTypeRule;
const VariablesDefaultValueAllowedRule: typeof PossibleFragmentSpreadsRule = (gql as any).VariablesDefaultValueAllowedRule;

import { RelayQLDefinition } from "./RelayQLAST";
import { Printable, Substitution } from "./RelayQLPrinter";
import {
  DocumentNode,
  GraphQLError,
  GraphQLFormattedError,
  GraphQLSchema,
} from "graphql";

export type Validator<T> = () => {
  validate: (schema: GraphQLSchema, ast: T) => Array<GraphQLError>,
};

type TransformerOptions = {
  inputArgumentName: string | null,
  snakeCase: boolean,
  substituteVariables: boolean,
  validator: Validator<any> | null,
};
type TextTransformOptions = {
  documentName: string,
  enableValidation: boolean,
  propName: string | null,
  tagName: string,
};

/**
 * Transforms a TemplateLiteral node into a RelayQLDefinition, which is then
 * transformed into a TS AST via RelayQLPrinter.
 */
export class RelayQLTransformer {
  schema: GraphQLSchema;
  options: TransformerOptions;

  constructor(schema: GraphQLSchema, options: TransformerOptions) {
    this.schema = schema;
    this.options = options;
  }

  transform(
    node: ts.TaggedTemplateExpression,
    options: TextTransformOptions,
  ): Printable {
    const {
    substitutions,
      templateText,
      variableNames,
  } = this.processTemplateLiteral(node, options.documentName);
    const documentText = this.processTemplateText(templateText, options);
    const definition = this.processDocumentText(documentText, options);

    const Printer = RelayQLPrinter(this.options);
    return new Printer(options.tagName, variableNames).print(
      definition,
      substitutions,
      options.enableValidation,
    );
  }

  /**
   * Convert TemplateLiteral into a single template string with substitution
   * names, a matching array of substituted values, and a set of substituted
   * variable names.
   */
  processTemplateLiteral(
    node: ts.TaggedTemplateExpression,
    documentName: string,
  ): {
      substitutions: Array<Substitution>,
      templateText: string,
      variableNames: { [variableName: string]: void },
    } {
    const chunks: string[] = [];
    const variableNames: { [key: string]: undefined } = {};
    const substitutions: Substitution[] = [];
    const template = node.template;
    if (ts.isNoSubstitutionTemplateLiteral(template)) {
      return {
        substitutions, templateText: template.text, variableNames,
      };
    }
    chunks.push(template.head.text);
    template.templateSpans.forEach((element, ii) => {
      const literal = element.literal;
      const chunk = literal.text;
      chunks.push(chunk);
      const name = 'RQL_' + ii;
      const value = element.expression;
      substitutions.push({ name, value });
      if (/:\s*$/.test(chunk)) {
        if (!this.options.substituteVariables) {
          throw new Error(util.format(
            "You supplied a GraphQL document named `%s` that uses template " +
            "substitution for an argument value, but variable substitution " +
            "has not been enabled.",
            documentName,
          ));
        }
        chunks.push('$' + name);
        variableNames[name] = undefined;
      } else {
        chunks.push('...' + name);
      }
    });
    return { substitutions, templateText: chunks.join('').trim(), variableNames };
  }

  /**
   * Converts the template string into a valid GraphQL document string.
   */
  processTemplateText(
    templateText: string,
    { documentName, propName }: TextTransformOptions,
  ): string {
    const pattern = /^(fragment|mutation|query|subscription)\s*(\w*)?([\s\S]*)/;
    const matches = pattern.exec(templateText);
    if (!matches) {
      throw new Error(util.format(
        "You supplied a GraphQL document named `%s` with invalid syntax. It " +
        "must start with `fragment`, `mutation`, `query`, or `subscription`.",
        documentName,
      ));
    }
    const type = matches[1];
    let name = matches[2] || documentName;
    let rest = matches[3];
    // Allow `fragment on Type {...}`.
    if (type === 'fragment' && name === 'on') {
      name =
        documentName + (propName ? '_' + capitalize(propName) : '') + 'RelayQL';
      rest = 'on' + rest;
    }
    const definitionName = capitalize(name);
    return type + ' ' + definitionName + ' ' + rest;
  }

  /**
   * Parses the GraphQL document string into a RelayQLDocument.
   */
  processDocumentText(
    documentText: string,
    { documentName, enableValidation }: TextTransformOptions,
  ): RelayQLDefinition<any> {
    const document = parse(new Source(documentText, documentName));
    const validationErrors = enableValidation
      ? this.validateDocument(document, documentName)
      : null;
    if (validationErrors) {
      const error = new Error(
        util.format(
          'You supplied a GraphQL document named `%s` with validation errors.',
          documentName,
        ),
      );
      (error as any).validationErrors = validationErrors;
      (error as any).sourceText = documentText;
      throw error;
    }
    const definition = document.definitions[0];

    const context = {
      definitionName: capitalize(documentName),
      isPattern: false,
      generateID: createIDGenerator(),
      schema: this.schema,
    };
    if (definition.kind === 'FragmentDefinition') {
      return new RelayQLFragment(context, definition);
    } else if (definition.kind === 'OperationDefinition') {
      if (definition.operation === 'mutation') {
        return new RelayQLMutation(context, definition);
      } else if (definition.operation === 'query') {
        return new RelayQLQuery(context, definition);
      } else if (definition.operation === 'subscription') {
        return new RelayQLSubscription(context, definition);
      } else {
        throw new Error(util.format("Unsupported operation: %s", definition.operation));
      }
    } else {
      throw new Error(util.format("Unsupported definition: %s", definition.kind));
    }
  }

  validateDocument(
    document: DocumentNode,
    documentName: string,
  ): Array<GraphQLFormattedError> | null {
    if (document.definitions.length !== 1) {
      throw new Error(util.format(
        "You supplied a GraphQL document named `%s` with %d definitions, but " +
        "it must have exactly one definition.",
        documentName,
        document.definitions.length,
      ));
    }
    const definition = document.definitions[0];
    const isMutation =
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'mutation';

    const validator = this.options.validator;
    let validationErrors;
    if (validator) {
      validationErrors = validator().validate(this.schema, document);
    } else {
      const rules = [
        FieldsOnCorrectTypeRule,
        FragmentsOnCompositeTypesRule,
        KnownArgumentNamesRule,
        KnownTypeNamesRule,
        PossibleFragmentSpreadsRule,
        ValuesOfCorrectTypeRule,
        VariablesDefaultValueAllowedRule,
        VariablesInAllowedPositionRule,
      ];
      if (!isMutation) {
        rules.push(ProvidedNonNullArgumentsRule);
      }
      validationErrors = validate(this.schema, document, rules);
    }

    if (validationErrors && validationErrors.length > 0) {
      return validationErrors.map(formatError);
    }
    return null;
  }
}

function capitalize(string: string): string {
  return string[0].toUpperCase() + string.slice(1);
}

/**
 * Utility to generate locally scoped auto-incrementing IDs.
 */
function createIDGenerator(): () => string {
  let _id = 0;
  return () => (_id++).toString(32);
}
