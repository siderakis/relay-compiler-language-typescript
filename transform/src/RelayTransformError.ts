import { Location } from "graphql";

export class RelayTransformError {
  message: string;
  loc: Location | null;
  stack: string;

  constructor(message: string, loc: Location | undefined) {
    this.message = message;
    this.loc = loc || null;
    this.stack = new Error().stack || "";
  }
}
