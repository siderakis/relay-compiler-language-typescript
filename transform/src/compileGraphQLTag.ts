// const createClassicNode = require('./createClassicNode');
// const createCompatNode = require('./createCompatNode');
import { createModernNode } from "./createModernNode";
import { getFragmentNameParts } from "./getFragmentNameParts";
import { DocumentNode, FragmentDefinitionNode, OperationDefinitionNode } from "graphql";
import * as ts from 'typescript';
import { NormalizedOptions } from "./Options";
import { createCompatNode } from "./createCompatNode";
import { createClassicNode } from "./createClassicNode";
import { bindingsAtNode } from "./bindingsAtNode";
import { setSourceMapRange } from "typescript";

/**
 * Given a graphql`` tagged template literal, replace it with the appropriate
 * runtime artifact.
 */
export function compileGraphQLTag(
  ctx: ts.TransformationContext,
  opts: NormalizedOptions,
  node: ts.TaggedTemplateExpression,
  ast: DocumentNode,
  fileName: string,
): ts.Expression {
  const mainDefinition = ast.definitions[0];

  if (mainDefinition.kind === 'FragmentDefinition') {
    const objPropName = getAssignedObjectPropertyName(node);
    if (objPropName) {
      if (ast.definitions.length !== 1) {
        throw new Error(
          'TSTransformRelay: Expected exactly one fragment in the ' +
          `graphql tag referenced by the property ${objPropName}.`,
        );
      }
      return createAST(ctx, opts, node, mainDefinition, fileName, true);
    }

    const nodeMap: { [key: string]: ts.Expression } = {};
    for (const definition of ast.definitions) {
      if (definition.kind !== 'FragmentDefinition') {
        throw new Error(
          'TSTransformRelay: Expected only fragments within this ' +
          'graphql tag.',
        );
      }

      const [, propName] = getFragmentNameParts(definition.name.value);
      nodeMap[propName] = createAST(ctx, opts, node, definition, fileName, false);
    }
    return createObject(nodeMap, node);
  }

  if (mainDefinition.kind === 'OperationDefinition') {
    if (ast.definitions.length !== 1) {
      throw new Error(
        'TSTransformRelay: Expected exactly one operation ' +
        '(query, mutation, or subscription) per graphql tag.',
      );
    }
    return createAST(ctx, opts, node, mainDefinition, fileName, true);
  }

  throw new Error(
    'TSTransformRelay: Expected a fragment, mutation, query, or ' +
    'subscription, got `' +
    mainDefinition.kind +
    '`.',
  );
}

function createAST(
  ctx: ts.TransformationContext,
  opts: NormalizedOptions,
  node: ts.TaggedTemplateExpression,
  graphqlDefinition: FragmentDefinitionNode | OperationDefinitionNode,
  fileName: string,
  setSoueceMapRange: boolean,
) {
  const isCompatMode = Boolean(opts.compat);
  const isDevVariable = opts.isDevVariable;
  const artifactDirectory = opts.artifactDirectory;
  const buildCommand =
    (opts.buildCommand) || 'relay-compiler';

  // Fallback is 'true'
  const isDevelopment =
    (process.env.NODE_ENV) !== 'production';

  const modernNode = createModernNode(ctx, opts, graphqlDefinition, fileName);
  if (isCompatMode) {
    console.log('Launching compat mode!');
    const result = createCompatNode(
      modernNode,
      createClassicNode(ctx, bindingsAtNode(node), node, graphqlDefinition, opts),
    );
    console.log('Done generating stuff');
    if (setSourceMapRange) {
      ts.setSourceMapRange(result, ts.getSourceMapRange(node));
    }
    return result;
  }
  if (setSourceMapRange) {
    ts.setSourceMapRange(modernNode, ts.getSourceMapRange(node));
  }
  return modernNode;
}

const idRegex = /^[$a-zA-Z_][$a-z0-9A-Z_]*$/;

function createObject(obj: { [propName: string]: ts.Expression }, originalNode: ts.Node) {
  const propNames = Object.keys(obj);

  const assignments = propNames.map(propName => {
    const name = idRegex.test(propName) ? ts.createIdentifier(propName) : ts.createLiteral(propName);
    return ts.createPropertyAssignment(name, obj[propName])
  });

  const objectLiteralNode = ts.createObjectLiteral(assignments, /* multiLine */ true);
  ts.setSourceMapRange(objectLiteralNode, ts.getSourceMapRange(originalNode));
  return objectLiteralNode;
}

function getAssignedObjectPropertyName(node: ts.Node): string | undefined {
  if (node.parent == null) {
    return undefined;
  }

  if (!ts.isPropertyAssignment(node.parent)) {
    return undefined;
  }

  const propName = node.parent.name;

  if (ts.isIdentifier(propName)) {
    return propName.text;
  }
  if (ts.isStringLiteral(propName)) {
    return propName.text;
  }
  return undefined;
}
