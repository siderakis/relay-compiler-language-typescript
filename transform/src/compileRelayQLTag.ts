import { createTransformError } from "./createTransformError";
import * as ts from "typescript";
import { NormalizedOptions } from "./Options";
import { RelayQLTransformer } from "./RelayQLTransformer";

/**
 * Given all the metadata about a found RelayQL tag, compile it and return
 * the resulting TS AST.
 */
export function compileRelayQLTag(
	ctx: ts.TransformationContext,
	options: NormalizedOptions,
	transformer: RelayQLTransformer,
	node: ts.TaggedTemplateExpression,
	documentName: string,
	propName: string | null,
	tagName: string,
	enableValidation: boolean,
): ts.Expression {
	const result = transformer.transform(node, {
		documentName,
		propName,
		tagName,
		enableValidation,
	});
	ts.setSourceMapRange(result, ts.getSourceMapRange(node));
	return result;
}
