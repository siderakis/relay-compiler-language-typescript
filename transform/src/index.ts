import * as ts from 'typescript';
import { parse } from 'graphql';
import * as fs from 'fs';
import { getValidGraphQLTag } from './getValidGraphQLTag';
import { NormalizedOptions, normalizeOptions, Options } from './Options';
import { compileGraphQLTag } from './compileGraphQLTag';

// https://github.com/Microsoft/TypeScript/blob/cc6d18e4db924d05e55c2a22587ad47ba53e7989/src/compiler/types.ts#L4490
const enum TransformFlags {
  ContainsES2015 = 1 << 7
}

interface ExtraNode {
  transformFlags: TransformFlags;
}

function visitor(ctx: ts.TransformationContext, sf: ts.SourceFile, opts: NormalizedOptions): ts.Visitor {
  const fileName = sf.fileName;
  const visit = (node: ts.Node): ts.Node => {
    // Easy bailout if there are not ES2015 features used
    if (((node as any as ExtraNode).transformFlags & TransformFlags.ContainsES2015) !== TransformFlags.ContainsES2015) {
      return node;
    }

    if (ts.isTaggedTemplateExpression(node)) {
      const ast = getValidGraphQLTag(node);
      if (ast) {
        return compileGraphQLTag(ctx, opts, node, ast, fileName);
      }
    }
    return ts.visitEachChild(node, visit, ctx);
  };

  return visit;
}

export function transformer(opts?: Options) {
  const options = normalizeOptions(opts || {});
  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
    return (sf: ts.SourceFile): ts.SourceFile => ts.visitNode(sf, visitor(context, sf, options));
  }
}
