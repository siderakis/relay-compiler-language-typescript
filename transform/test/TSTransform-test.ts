import { transformer } from '../src';
import * as ts from 'typescript';
import * as path from 'path';
import { Options } from '../src/Options';

function transformWithOptions(options: Options, contents: string, fileName: string): ts.TranspileOutput {
  return ts.transpileModule(contents, {
    compilerOptions: {
      target: ts.ScriptTarget.ES5,
      sourceMap: true,
    },
    fileName: fileName,
    transformers: {
      before: [transformer(options)],
    },
  });
}

const schemaPath = path.resolve(__dirname, 'testschema.rfc.graphql');

describe('TSTransform', () => {
  it('Modern should compile', async () => {
    const text = 'createFragmentContainer(MyComponent, {todo: graphql`fragment MyFragment_todo on MyType { id }`})';

    expect(transformWithOptions({ isDevVariable: 'IS_DEV', artifactDirectory: '/testing/artifacts' }, text, '/test/MyComponent.ts')).toMatchSnapshot('modern test');
  });
  it('Classic should compile', async () => {
    const text = 'createFragmentContainer(MyComponent, {todo: () => Relay.QL`fragment on Node { id }`})';

    expect(transformWithOptions({ isDevVariable: 'IS_DEV', artifactDirectory: '/testing/artifacts', schema: schemaPath }, text, '/test/MyComponent.ts')).toMatchSnapshot('classic test');
  });
});
