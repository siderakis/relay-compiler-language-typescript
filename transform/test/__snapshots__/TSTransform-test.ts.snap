// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSTransform Classic should compile: classic test 1`] = `
Object {
  "diagnostics": Array [],
  "outputText": "createFragmentContainer(MyComponent, { todo: function () { return function () {
        return {
            children: [
                {
                    fieldName: \\"id\\",
                    kind: \\"Field\\",
                    metadata: {
                        isRequisite: true
                    },
                    type: \\"String\\"
                },
                {
                    fieldName: \\"__typename\\",
                    kind: \\"Field\\",
                    metadata: {
                        isGenerated: true,
                        isRequisite: true
                    },
                    type: \\"String\\"
                }
            ],
            id: Relay.QL.__id(),
            kind: \\"Fragment\\",
            metadata: {
                isAbstract: true
            },
            name: \\"MyComponent_TodoRelayQL\\",
            type: \\"Node\\"
        };
    }(); } });
//# sourceMappingURL=MyComponent.js.map",
  "sourceMapText": "{\\"version\\":3,\\"file\\":\\"MyComponent.js\\",\\"sourceRoot\\":\\"\\",\\"sources\\":[\\"MyComponent.ts\\"],\\"names\\":[],\\"mappings\\":\\"AAAA,uBAAuB,CAAC,WAAW,EAAE,EAAC,IAAI,EAAE,qBAAM;;;;;;;;;;;;;;;;;;;;;;;;;;;;;OAAiC,GAAA,EAAC,CAAC,CAAA\\"}",
}
`;

exports[`TSTransform Modern should compile: modern test 1`] = `
Object {
  "diagnostics": Array [],
  "outputText": "createFragmentContainer(MyComponent, { todo: function () {
        if (IS_DEV) {
            var node = require(\\"../testing/artifacts/MyFragment_todo.graphql\\").default;
            if (node.hash && node.hash !== \\"e655cc161574f3976cf1b4a8310f6885\\") {
                console.error(\\"The definition of 'MyFragment_todo' appears to have changed. Run \`relay-compiler\` to update the generated files to receive the expected data.\\");
            }
        }
        return require(\\"../testing/artifacts/MyFragment_todo.graphql\\").default;
    } });
//# sourceMappingURL=MyComponent.js.map",
  "sourceMapText": "{\\"version\\":3,\\"file\\":\\"MyComponent.js\\",\\"sourceRoot\\":\\"\\",\\"sources\\":[\\"MyComponent.ts\\"],\\"names\\":[],\\"mappings\\":\\"AAAA,uBAAuB,CAAC,WAAW,EAAE,EAAC,IAAI,EAAE;;;;;;;;KAAkD,EAAC,CAAC,CAAA\\"}",
}
`;
